[flaky] test/cache-interceptor/cache-tests.mjs

```console
> node test/cache-interceptor/cache-tests.mjs --ci

Testing 4 environments

PROTOCOL: http

TEST ENVIRONMENT
       BASE_URL: http://localhost:8000
          store: MemoryCacheStore
        methods: ["GET","HEAD"]
           type: shared
  ignored tests: ["conditional-etag-forward-unquoted","conditional-etag-strong-generate-unquoted","cc-resp-no-cache","cc-resp-no-cache-case-insensitive","304-etag-update-response-Cache-Control","304-etag-update-response-Content-Foo","304-etag-update-response-Test-Header","304-etag-update-response-X-Content-Foo","304-etag-update-response-X-Test-Header","age-parse-float","head-200-update","head-200-retain","head-410-update","stale-close-must-revalidate","stale-close-no-cache","freshness-max-age-s-maxage-private","freshness-max-age-s-maxage-private-multiple"]
Listening on http://[::]:8000/
SERVER WARNING: State not found for 61a81b88-d55e-4c93-949b-96a13600148e

TEST ENVIRONMENT
       BASE_URL: http://localhost:8001
          store: SqliteCacheStore
        methods: ["GET","HEAD"]
           type: shared
  ignored tests: ["conditional-etag-forward-unquoted","conditional-etag-strong-generate-unquoted","cc-resp-no-cache","cc-resp-no-cache-case-insensitive","304-etag-update-response-Cache-Control","304-etag-update-response-Content-Foo","304-etag-update-response-Test-Header","304-etag-update-response-X-Content-Foo","304-etag-update-response-X-Test-Header","age-parse-float","head-200-update","head-200-retain","head-410-update","stale-close-must-revalidate","stale-close-no-cache","freshness-max-age-s-maxage-private","freshness-max-age-s-maxage-private-multiple"]
Listening on http://[::]:8001/
SERVER WARNING: State not found for 164e4435-3136-4b63-b399-a800fbe800c4

        Total tests: 283
            Skipped: 0 (0.0%)
             Passed: 219 (77.4%)
             Failed: 0 (0.0%)
  Failed (optional): 59 (20.8%)
       Setup failed: 5 (1.8%)
Test Harness Failed: 0 (0.0%)
  Dependency Failed: 0 (0.0%)
            Retried: 0 (0.0%)

TEST ENVIRONMENT
       BASE_URL: http://localhost:8002
          store: MemoryCacheStore
        methods: ["GET","HEAD"]
           type: private
  ignored tests: ["conditional-etag-forward-unquoted","conditional-etag-strong-generate-unquoted","cc-resp-no-cache","cc-resp-no-cache-case-insensitive","304-etag-update-response-Cache-Control","304-etag-update-response-Content-Foo","304-etag-update-response-Test-Header","304-etag-update-response-X-Content-Foo","304-etag-update-response-X-Test-Header","age-parse-float","head-200-update","head-200-retain","head-410-update","stale-close-must-revalidate","stale-close-no-cache","freshness-max-age-s-maxage-private","freshness-max-age-s-maxage-private-multiple"]
Listening on http://[::]:8002/
SERVER WARNING: State not found for bb0a8e8e-0170-4e93-95ed-7a51cc32018e

        Total tests: 283
            Skipped: 0 (0.0%)
             Passed: 220 (77.7%)
             Failed: 0 (0.0%)
  Failed (optional): 58 (20.5%)
       Setup failed: 5 (1.8%)
Test Harness Failed: 0 (0.0%)
  Dependency Failed: 0 (0.0%)
            Retried: 0 (0.0%)

TEST ENVIRONMENT
       BASE_URL: http://localhost:8003
          store: SqliteCacheStore
        methods: ["GET","HEAD"]
           type: private
  ignored tests: ["conditional-etag-forward-unquoted","conditional-etag-strong-generate-unquoted","cc-resp-no-cache","cc-resp-no-cache-case-insensitive","304-etag-update-response-Cache-Control","304-etag-update-response-Content-Foo","304-etag-update-response-Test-Header","304-etag-update-response-X-Content-Foo","304-etag-update-response-X-Test-Header","age-parse-float","head-200-update","head-200-retain","head-410-update","stale-close-must-revalidate","stale-close-no-cache","freshness-max-age-s-maxage-private","freshness-max-age-s-maxage-private-multiple"]
Listening on http://[::]:8003/

Error: Process completed with exit code 1.
```

Example: https://github.com/nodejs/undici/actions/runs/25363470627/job/74368317437

--- a/test/cache-interceptor/cache-tests-worker.mjs
+++ b/test/cache-interceptor/cache-tests-worker.mjs
@@ -2,6 +2,8 @@
 
 import { styleText } from 'node:util'
 import { exit } from 'node:process'
+import net from 'node:net'
+import { setTimeout as sleep } from 'node:timers/promises'
 import { getResults, runTests as runTestSuite } from '../fixtures/cache-tests/test-engine/client/runner.mjs'
 import { determineTestResult, testLookup } from '../fixtures/cache-tests/test-engine/lib/results.mjs'
 import tests from '../fixtures/cache-tests/tests/index.mjs'
@@ -27,6 +29,9 @@ if (environment.cacheStore) {
 
 // Start the test server
 await import('../fixtures/cache-tests/test-engine/server/server.mjs')
+// TODO: Remove when server.mjs waits for the server to listen.
+// https://github.com/http-tests/cache-tests/issues/161
+await waitForServer(process.env.BASE_URL)
 
 // Output the testing setup
 console.log('TEST ENVIRONMENT')
@@ -84,6 +89,57 @@ async function makeCacheStore (type) {
   }
 
   return new Store()
+}
+
+/**
+ * @param {string} baseUrl
+ * @returns {Promise<void>}
+ */
+async function waitForServer (baseUrl) {
+  const { hostname, port, protocol } = new URL(baseUrl)
+  const deadline = Date.now() + 30_000
+
+  while (Date.now() < deadline) {
+    if (await canConnect(hostname, Number(port), protocol)) {
+      return
+    }
+
+    await sleep(250)
+  }
+
+  throw new Error(`cache test server did not start listening at ${baseUrl}`)
+}
+
+/**
+ * @param {string} host
+ * @param {number} port
+ * @param {string} protocol
+ * @returns {Promise<boolean>}
+ */
+function canConnect (host, port, protocol) {
+  const { promise, resolve } = Promise.withResolvers()
+  const socket = net.connect({ host, port })
+
+  socket.once('connect', () => {
+    socket.end()
+    resolve(true)
+  })
+
+  socket.once('error', () => {
+    resolve(false)
+  })
+
+  socket.setTimeout(1_000, () => {
+    socket.destroy()
+    resolve(false)
+  })
+
+  if (protocol !== 'http:' && protocol !== 'https:') {
+    socket.destroy()
+    resolve(false)
+  }
+
+  return promise
 }
 
 /**

